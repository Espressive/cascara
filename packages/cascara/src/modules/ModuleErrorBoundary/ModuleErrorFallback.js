import React from 'react';

import styles from '../DataModule.module.scss';

const ModuleErrorFallback = () => {
  return (
    <div className={styles.Error} role='alert'>
      Error
    </div>
  );
};

export default ModuleErrorFallback;
