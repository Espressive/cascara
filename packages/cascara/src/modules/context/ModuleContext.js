import { createContext } from 'react';

// We should ONLY put values here that are needed for ALL Modules
const defaultValue = {
  isEditing: false,
  isEditingAny: false,
  value: 'blue',
};

/*
 * We will use this baseline context to create any of our context providers from.
 * Any unique functionality should be included in that provider and not in this Context.
 */
const ModuleContext = createContext(defaultValue);

/*
 * We are also exporting the default value in case we need to hydrate or merge
 * any providers with these default values.
 */
export { defaultValue };

export default ModuleContext;
