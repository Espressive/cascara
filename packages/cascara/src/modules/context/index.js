export { default as ModuleContext } from './ModuleContext';
export { default as ModuleProvider } from './ModuleProvider';
