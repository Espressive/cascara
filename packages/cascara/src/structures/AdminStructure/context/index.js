export { default as AdminContext } from './AdminContext';
export { default as AdminProvider } from './AdminProvider';
