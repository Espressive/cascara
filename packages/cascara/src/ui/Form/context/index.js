import FormProvider from './FormProvider';

export { FormProvider };
