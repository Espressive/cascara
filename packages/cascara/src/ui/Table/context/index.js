export * from './TableContext';
export { default } from './TableContext';
