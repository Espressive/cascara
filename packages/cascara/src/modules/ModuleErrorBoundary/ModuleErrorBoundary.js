import React from 'react';
import pt from 'prop-types';

import { Boundaries } from '../../system-components';
import ModuleErrorFallback from './ModuleErrorFallback';

const propTypes = {
  children: pt.oneOfType([pt.arrayOf(pt.node), pt.node]),
};

const ModuleErrorBoundary = ({ children }) => {
  return (
    <Boundaries ErrorComponent={ModuleErrorFallback}>{children}</Boundaries>
  );
};

ModuleErrorBoundary.propTypes = propTypes;

export default ModuleErrorBoundary;
