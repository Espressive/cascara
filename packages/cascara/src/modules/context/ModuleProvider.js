import React from 'react';
import pt from 'prop-types';

import ModuleContext, { defaultValue } from './ModuleContext';
import useToggle from '../../hooks/useToggle';

const propTypes = {
  children: pt.oneOfType([pt.element, pt.arrayOf(pt.element)]),
  value: pt.shape({
    isEditing: pt.bool,
  }),
};

const ModuleProvider = ({ children, value, ...props }) => {
  const [isEditing, setIsEditing] = useToggle(value?.isEditing || false);

  const mergedValues = {
    ...defaultValue,
    isEditing,
    setIsEditing,
    ...value,
  };

  return (
    <ModuleContext.Provider value={mergedValues} {...props}>
      {children}
    </ModuleContext.Provider>
  );
};

ModuleProvider.propTypes = propTypes;

export default ModuleProvider;
