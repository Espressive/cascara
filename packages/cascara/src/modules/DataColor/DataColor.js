import React, { useContext } from 'react';
import { Role } from 'reakit/Role';
import { Input } from 'reakit/Input';
import pt from 'prop-types';
import { ModuleContext } from '../context';
import styles from '../DataModule.module.scss';

import ModuleErrorBoundary from '../ModuleErrorBoundary';
import getAccessibleLabelSetters from '../helpers';

const propTypes = {
  /** A module can have an Attribute, which will be used as form field name */
  attribute: pt.string,
  /** A Module can be defined to not present an editing state */
  isEditable: pt.bool,
  /** Presents the input without a label. NOT USER CONFIGURABLE */
  isLabeled: pt.bool,
  /** A Module needs to have a unique label relative to its context */
  label: pt.string,
  /** A Module can have a value */
  value: pt.string,
};

const DataColor = ({
  attribute,
  isEditable = true,
  isLabeled = true,
  label,
  value,
  ...rest
}) => {
  const { isEditing, formMethods } = useContext(ModuleContext);
  const { setAriaLabel, setHtmlFor } = getAccessibleLabelSetters(
    isLabeled,
    label
  );

  const renderEditing = (
    <label htmlFor={setHtmlFor}>
      {label && isLabeled && <span className={styles.LabelText}>{label}</span>}
      <Input
        {...rest}
        aria-label={setAriaLabel}
        className={styles.Color}
        defaultValue={value}
        id={label}
        name={attribute || label}
        ref={formMethods?.register}
        type='color'
      />
    </label>
  );

  const renderDisplay = (
    <span>
      {label && isLabeled && <span className={styles.LabelText}>{label}</span>}
      <span aria-label={label} className={styles.Color} {...rest}>
        {value}
      </span>
    </span>
  );

  // Do not render an editable input if the module is not editable
  return (
    <ModuleErrorBoundary>
      <Role className={styles.Color}>
        {isEditing && isEditable ? renderEditing : renderDisplay}
      </Role>
    </ModuleErrorBoundary>
  );
};

DataColor.propTypes = propTypes;

export default DataColor;
