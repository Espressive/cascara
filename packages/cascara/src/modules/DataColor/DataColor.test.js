import { fireEvent, render, screen, waitFor } from '@testing-library/react';

import cosmosFixtures, {
  displayProps,
  editingProps,
} from './DataColor.fixture';
import testFixtures from '../helpers/ModuleTestHelper.fixture';

const { display, editing, displayNoLabel, editingNoLabel } = cosmosFixtures;
const { colorTest } = testFixtures;

describe('DataColor', () => {
  // without ModuleSandbox will render the property information into a span

  describe('display', () => {
    // We need a place to store the view for snapshot testing. This is not required when we are using `screen` directly from RTL.
    let view;

    beforeEach(() => {
      // Set the render container to our `view` so it is in scope for the snapshot test
      view = render(display).container;
    });

    test('snapshot', () => {
      expect(view).toMatchSnapshot();
    });

    test('renders a <span> by default', () => {
      const input = screen.getByLabelText(displayProps.label);
      // Make sure the actual DOM element is not render an input
      expect(input.tagName).toMatch('SPAN');
      // Make sure the dom element that has our aria-label is the input
      expect(input.classList.contains('Color')).toBe(true);
    });
  });

  describe('editing', () => {
    // We need a place to store the view for snapshot testing. This is not required when we are using `screen` directly from RTL.
    let view;

    beforeEach(() => {
      // Set the render container to our `view` so it is in scope for the snapshot test
      view = render(editing).container;
    });

    test('snapshot', () => {
      expect(view).toMatchSnapshot();
    });

    test('renders a <input email> by default', () => {
      const input = screen.getByLabelText(editingProps.label);
      // Check that we also use the correct type
      expect(input).toHaveAttribute('type', 'color');
    });

    test('change value', () => {
      const newColor = '#e66465';
      const input = screen.getByLabelText(editingProps.label);
      fireEvent.change(input, { target: { value: newColor } });
      expect(input).toHaveValue(newColor);
    });
  });

  describe('accessibility', () => {
    test('editing', () => {
      render(editing);

      const input = screen.getByLabelText(editingProps.label);
      // The label tag is the parent wrapper
      const label = input.closest('label');

      // Test is written this way to make sure we know that both values need to be the same.
      const linkedLabelValue = editingProps.label;

      // Verify label for attribute has linked value
      expect(label).toHaveAttribute(
        'for',
        expect.stringContaining(linkedLabelValue)
      );
      // Verify input id attribute has linked value
      expect(input).toHaveAttribute(
        'id',
        expect.stringContaining(linkedLabelValue)
      );
      // Check that the input does NOT have an aria-label defined because there is a label tag
      expect(input).not.toHaveAttribute('aria-label');
    });

    test('display no label', () => {
      // Make sure that the input is still accessible with label text even when we are not showing a label tag in tables
      render(displayNoLabel);
      const input = screen.getByLabelText(displayProps.label);
      expect(input).toBeDefined();
    });

    test('editing no label', () => {
      // Make sure that the input is still accessible with label text even when we are not showing a label tag in tables
      render(editingNoLabel);
      const input = screen.getByLabelText(editingProps.label);
      expect(input).toBeDefined();
    });
  });

  describe('data types', () => {
    /**
     * This test validates the data types going in are also going out.
     * i.e. without any type casting that corrupts the data.
     *
     * The test focuses on the data that is NOT mutated.
     */
    test("what goes in, must go out - 'at rest' data", async () => {
      // prepare action handler mock
      const onAction = jest.spyOn(console, 'log');

      render(colorTest);

      // wait for checkbox to be present
      await waitFor(() => screen.findByRole('checkbox', { name: 'Checkbox' }));

      // get the checkbox and check it
      const checkboxInput = screen.getByRole('checkbox', { name: 'Checkbox' });
      fireEvent.change(checkboxInput, { target: { checked: true } });

      // submit the form
      const saveButton = screen.getByRole('button', { name: 'Save' });
      fireEvent.click(saveButton);

      // wait for the form to be in display mode
      await waitFor(() => screen.findByRole('button', { name: 'Edit' }));

      /**
       * We expect the mock to have been called with two params:
       *
       * - the action object which includes the action name
       * - the data sibmitted by the form
       */
      expect(onAction).toBeCalledWith(
        { name: 'edit.save' },
        {
          checkbox: true,
          color: '#a1a1a1',
        }
      );

      /**
       * We expect that outbound data type is the same as
       * the inblound data type.
       */
      const submittedValue = onAction.mock.calls[0][1].color;
      const submittedValueType = typeof submittedValue;
      expect(submittedValueType).toBe('string');
    });

    /**
     * This test validates the data types going in are also going out.
     * i.e. without any type casting that corrupts the data.
     *
     * The test focuses on the data that is mutated.
     */
    test("what goes in, must go out - 'mutated' data", async () => {
      // prepare action handler mock
      const onAction = jest.spyOn(console, 'log');

      render(colorTest);

      // wait for input to be present
      await waitFor(() => screen.getByRole('checkbox', { name: 'Checkbox' }));

      // make the form dirty
      const colorInput = screen.getByLabelText('Color');
      fireEvent.input(colorInput, { target: { value: '#e1e1e1' } });

      // submit the form
      const saveButton = screen.getByRole('button', { name: 'Save' });
      fireEvent.click(saveButton);

      // wait for the form to be in display mode
      await waitFor(() => screen.findByRole('button', { name: 'Edit' }));

      /**
       * We expect the mock to have been called with two params:
       *
       * - the action object which includes the action name
       * - the data sibmitted by the form
       */
      expect(onAction).toBeCalledWith(
        { name: 'edit.save' },
        {
          checkbox: false,
          color: '#e1e1e1',
        }
      );

      /**
       * We expect that outbound data type is the same as
       * the inblound data type.
       */
      const submittedValue = onAction.mock.calls[0][1].color;
      const submittedValueType = typeof submittedValue;
      expect(submittedValueType).toBe('string');
    });
  });
});
