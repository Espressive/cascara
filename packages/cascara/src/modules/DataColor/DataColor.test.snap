// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataColor display snapshot 1`] = `
<div>
  <div
    class="Color"
  >
    <span>
      <span
        class="LabelText"
      >
        Display
      </span>
      <span
        aria-label="Display"
        class="Color"
      >
        #f6b73c
      </span>
    </span>
  </div>
</div>
`;

exports[`DataColor editing snapshot 1`] = `
<div>
  <div
    class="Color"
  >
    <label
      for="Editing"
    >
      <span
        class="LabelText"
      >
        Editing
      </span>
      <input
        class="Color"
        id="Editing"
        name="Editing"
        type="color"
        value="#f6b73c"
      />
    </label>
  </div>
</div>
`;
